server/src/main/java/org/elasticsearch/cluster/metadata/SystemIndexMetadataUpgradeService.java

```diff
@@ -78,9 +84,17 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                 if (cursor.value != lastIndexMetadataMap.get(cursor.key)) {
                     final boolean isSystem = systemIndices.isSystemIndex(cursor.value.getIndex()) ||
                         systemIndices.isSystemIndexBackingDataStream(cursor.value.getIndex().getName());
+                    IndexMetadata.Builder builder = IndexMetadata.builder(cursor.value);
                     if (isSystem != cursor.value.isSystem()) {
-                        updatedMetadata.add(IndexMetadata.builder(cursor.value).system(cursor.value.isSystem() == false).build());
+                        builder.system(cursor.value.isSystem() == false);
+                        if (isSystem && cursor.value.getSettings().getAsBoolean(IndexMetadata.SETTING_INDEX_HIDDEN, false)) {
```

Comment:
this should not be nested within the if statement and should be its own check after the system flag comparison

Reply:
That makes sense. I was thinking that the check would only be needed when we are changing an index anyway, but you're right that it should be a general operation in any case where the service finds a "hidden system index."